File: core/src/main/scala/eu/easyminer/rdf/index/TripleItemHashIndex.scala
package eu.easyminer.rdf.index

import eu.easyminer.rdf.data.TripleItem

/**
  * Created by Vaclav Zeman on 12. 3. 2018.
  */
class TripleItemHashIndex private(map: collection.Map[Int, TripleItem]) {

  def getIndex(x: TripleItem): Int = Stream.iterate(x.hashCode())(_ + 1).find(i => map(i) == x).get

  def getTripleItem(x: Int): TripleItem = map(x)

  def iterator: Iterator[(Int, TripleItem)] = map.iterator

}

object TripleItemHashIndex {

  def fromIndexedItem(col: Traversable[(Int, TripleItem)]): TripleItemHashIndex = {
    new TripleItemHashIndex(collection.mutable.HashMap.empty[Int, TripleItem] ++= col)
  }

  def apply(col: Traversable[TripleItem]): TripleItemHashIndex = {
    val map = collection.mutable.HashMap.empty[Int, TripleItem]
    for (item <- col) {
      Stream.iterate(item.hashCode())(_ + 1).map(i => i -> map.get(i)).find(_._2.forall(_ == item)).get match {
        case (i, None) => map += (i -> item)
        case _ =>
      }
    }
    new TripleItemHashIndex(map)
  }

}

File: java/src/main/java/com/github/propi/rdfrules/java/index/TripleItemHashIndex.java
package com.github.propi.rdfrules.java.index;

import com.github.propi.rdfrules.java.TripleItemConverters;
import com.github.propi.rdfrules.java.data.TripleItem;

/**
 * Created by Vaclav Zeman on 11. 5. 2018.
 */
public class TripleItemHashIndex {

    final private com.github.propi.rdfrules.index.TripleItemHashIndex mapper;

    public TripleItemHashIndex(com.github.propi.rdfrules.index.TripleItemHashIndex mapper) {
        this.mapper = mapper;
    }

    public int getIndex(TripleItem tripleItem) {
        return mapper.getIndex(tripleItem.asScala());
    }

    public TripleItem getTripleItem(int index) {
        return TripleItemConverters.toJavaTripleItem(mapper.getTripleItem(index));
    }

}